### PR TITLE
fix potential hangers in AQL queries on server shutdown

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Fix potential hangers in server shutdown when there is a running AQL query

- [x] Bug-Fix for *devel-branch* (i.e. no need for backports?)
- [x] The behavior in this PR can be (and was) *manually tested* (support / qa / customers can test it)
- [x] The behavior change can be verified via automatic tests

### Testing & Verification

This change is already covered by existing tests, such as *cluster tests*.

http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/10317/

### CODE REVIEW

**Comment by jsteemann** [comment]
Tests blue

File: arangod/Aql/ExecutionEngine.cpp [hunk 1]
```diff
@@ -648,6 +648,13 @@ Result ExecutionEngine::shutdownSync(int errorCode) noexcept try {
         if (state == ExecutionState::WAITING) {
           sharedState->waitForAsyncWakeup();
         }
+        if ((errorCode == TRI_ERROR_INTERNAL || errorCode == TRI_ERROR_SHUTTING_DOWN) &&
```
Comment (mchacki) [inline arangod/Aql/ExecutionEngine.cpp:651]: From the comment below, I would expect this be added 3 lines above before the test for WAITING.

IF there is no one to wake us up, there is no point in waiting for someone.
If may be excluded by the behavior of waitForAsyncWakeup, but I would like to not depend on it here.

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/10397/

File: arangod/Aql/ExecutionEngine.cpp (hunk 1 above)
Comment (graetzer) [reply arangod/Aql/ExecutionEngine.cpp:649]: this seems unelegant, maybe throw in waitForAsyncWakeup if the shared state is not valid ?

**Comment by graetzer** [comment]
I think this is not elegant, I think you should throw if someone tries waiting for a invalid shared state

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/10450/

**Comment by jsteemann** [comment]
Tests blue

**Pull request merged**